trunk/source/ChcSolver/src/de/uni_freiburg/informatik/ultimate/plugins/chcsolver/ChcSolverObserver.java

```diff
@@ -0,0 +1,169 @@
+
+public class ChcSolverObserver extends BaseObserver {
+	private final IUltimateServiceProvider mServices;
+	private final ILogger mLogger;
+	private final ChcSolverPreferences mPrefs;
+
+	private ChcSolution mSolution;
+
+	public ChcSolverObserver(final IUltimateServiceProvider services, final ILogger logger,
+			final ChcSolverPreferences prefs) {
+		mServices = services;
+		mLogger = logger;
+		mPrefs = prefs;
+	}
+
+	@Override
+	public boolean process(final IElement root) throws Throwable {
+		if (!(root instanceof HornClauseAST)) {
+			return true;
+		}
+
+		final HornAnnot annot = HornAnnot.getAnnotation(root);
+		final IChcScript chcScript = getBackend(annot);
+		configureBackend(chcScript);
+
+		final var satisfiability = chcScript.solve(annot.getSymbolTable(), annot.getHornClauses());
+		final IResult result = createResult(chcScript, satisfiability);
+		mServices.getResultService().reportResult(Activator.PLUGIN_ID, result);
+
+		return false;
+	}
+
+	private IChcScript getBackend(final HornAnnot annotation) {
+		switch (mPrefs.getBackend()) {
+		case ELDARICA:
+			return new EldaricaChcScript(mServices, annotation.getScript().getScript());
+		case Z3:
+			// We use the script given in the annotation. For this to work, that script should use Z3.
+			// To use a fresh Z3 instance for solving instead, one has to transfer the Horn clause terms to that script.
+			return new SmtChcScript(annotation.getScript());
+		case TREEAUTOMIZER:
+			// NOTE: TAPreferences (last parameter) currently unused by TreeAutomizer
+			return new TreeAutomizerChcScript(mServices, annotation.getScript(), null);
+		case GOLEM:
+			return new GolemChcScript(mServices, annotation.getScript());
+		default:
+			throw new UnsupportedOperationException();
+		}
+	}
+
+	private void configureBackend(final IChcScript backend) {
+		if (backend.supportsModelProduction()) {
+			backend.produceModels(mPrefs.produceModels());
+		} else if (mPrefs.produceModels()) {
+			mLogger.warn("Model production is not supported by backend");
+		}
+
+		if (backend.supportsDerivationProduction()) {
+			backend.produceDerivations(mPrefs.produceDerivation());
+		} else if (mPrefs.produceDerivation()) {
+			mLogger.warn("Derivation production is not supported by backend");
+		}
+
+		if (backend.supportsUnsatCores()) {
+			backend.produceUnsatCores(mPrefs.produceUnsatCore());
+		} else if (mPrefs.produceUnsatCore()) {
+			mLogger.warn("UNSAT core production is not supported by backend");
+		}
+	}
+
+	private IResult createResult(final IChcScript chcScript, final LBool satisfiability) {
+		switch (satisfiability) {
+		case SAT:
+			return createSatResult(chcScript);
+		case UNSAT:
+			return createUnSatResult(chcScript);
+		case UNKNOWN:
+			mSolution = ChcSolution.unknown();
+			return new ChcUnknownResult(Activator.PLUGIN_ID, "CHC solver returned UNKNOWN.");
+		default:
+			throw new IllegalStateException();
```

Comment:
Better UnsupportedOperationException with the unsupported value (or AssertionError)